### PR TITLE
fix(RuleSet): normalize mixed use array and function

### LINKED ISSUE #3484: Crash with rule.use as a function returning array
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When the property `use` of a rule is defined as a function, it only works if it returns a single loader object. If it returns an array of loaders, it will crash with this error:
```
[...]/node_modules/enhanced-resolve/lib/Resolver.js:151
    var idxQuery = identifier.indexOf("?");
                             ^

TypeError: Cannot read property 'indexOf' of undefined
    at Resolver.parse ([...]/node_modules/enhanced-resolve/lib/Resolver.js:151:27)
    at Resolver.<anonymous> ([...]/node_modules/enhanced-resolve/lib/ParsePlugin.js:16:25)
    at Resolver.applyPluginsParallelBailResult1 ([...]/node_modules/tapable/lib/Tapable.js:252:14)
    at beforeInnerCallback ([...]/node_modules/enhanced-resolve/lib/Resolver.js:112:19)
    at loggingCallbackWrapper ([...]/node_modules/enhanced-resolve/lib/createInnerCallback.js:31:19)
    at Resolver.applyPluginsAsyncSeriesBailResult1 ([...]/node_modules/tapable/lib/Tapable.js:145:46)
    at Resolver.doResolve ([...]/node_modules/enhanced-resolve/lib/Resolver.js:100:11)
    at Resolver.<anonymous> ([...]/node_modules/enhanced-resolve/lib/UnsafeCachePlugin.js:36:12)
    at Resolver.applyPluginsParallelBailResult1 ([...]/node_modules/tapable/lib/Tapable.js:252:14)
    at beforeInnerCallback ([...]/node_modules/enhanced-resolve/lib/Resolver.js:112:19)
    at loggingCallbackWrapper ([...]/node_modules/enhanced-resolve/lib/createInnerCallback.js:31:19)
    at Resolver.applyPluginsAsyncSeriesBailResult1 ([...]/node_modules/tapable/lib/Tapable.js:145:46)
    at Resolver.doResolve ([...]/node_modules/enhanced-resolve/lib/Resolver.js:100:11)
    at Resolver.resolve ([...]/node_modules/enhanced-resolve/lib/Resolver.js:74:14)
    at [...]/node_modules/webpack/lib/NormalModuleFactory.js:252:12
    at [...]/node_modules/async/dist/async.js:1008:9
```

**If the current behavior is a bug, please provide the steps to reproduce.**
Just try some config with a `use` function returning array. Something like this:
```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.gif$/,
        use: function (data) {
          // if it returned a single loader, it wouldn't fail
          // return {loader: 'file-loader'};

          // returning a chain of loaders: will fail
          return [
            {loader: 'file-loader'},
            {loader: 'image-webpack-loader'}
          ];
        }
      }
    ]
  }
};
```

**What is the expected behavior?**
It should work returning an array of loaders for chaining, as it works defining a plain array to `use`.

**Please mention other relevant information such as the browser version, Node.js version, Operating System and programming language.**
Node.js: v7.2.1
Webpack: v2.2.0-rc.0

### PR DESCRIPTION
**Pull request by TheLarkInn**
**What kind of change does this PR introduce?**
Bugfix

 
**Did you add tests for your changes?**
Yes 

**If relevant, link to documentation update:**

**Summary**
Fixes #3484

**Does this PR introduce a breaking change?**
No

**Other information**

### CODE REVIEW

File: lib/RuleSet.js [hunk 1]
```diff
@@ -224,21 +224,32 @@ RuleSet.normalizeUse = function normalizeUse(use) {
 	return [RuleSet.normalizeUseItem(use)];
 };
 
+RuleSet.normalizeUseItemFunction = function normalizeUseItemFunction(use, data) {
+	if(typeof use(data) === "string") {
```
Comment (sokra) [inline lib/RuleSet.js]: don't call the `use` function multiple times. Cache the result.

**Comment by sokra** [comment]
It was also requested to have a use function return an array of use items:

``` js
rules: [
  {
    use: function(data) {
      return [
        { loader: "a-loader", options: "abc" },
        "b-loader"
      ]
    }
  }
]
```

**Comment by sokra** [comment]
let's merge this for now, the other change can go in a separate PR.

**Comment by twaddington** [comment]
This bit me today: I was trying to return an array from a use function as indicated in https://github.com/webpack/webpack/pull/3557#issuecomment-268958481.

I managed to get things working another way but this would have saved me a few hours. Any plan to fix this? We're on webpack `^3.3.0`.

**Comment by sokra** [comment]
@twaddington If you want to you can send a PR allowing this.

**Comment by twaddington** [comment]
> @twaddington If you want to you can send a PR allowing this. &mdash; @sokra 

I'll see if I can get that working. Just wanted to make sure it wasn't missing something.

**Comment by carmanchris31** [comment]
@twaddington I'm also trying to use a function to return array but I'm getting the error thrown from https://github.com/webpack/enhanced-resolve/blob/c69c626e585cdd46444be04d42c9e31ba99e3b80/lib/Resolver.js#L181

**Pull request merged**